File: shared/src/commonMain/kotlin/org/nihongo/mochi/domain/util/LevelContentProvider.kt
package org.nihongo.mochi.domain.util

import org.nihongo.mochi.data.ScoreManager
import org.nihongo.mochi.data.ScoreRepository
import org.nihongo.mochi.domain.kana.KanaRepository
import org.nihongo.mochi.domain.kana.KanaType
import org.nihongo.mochi.domain.kanji.KanjiRepository
import org.nihongo.mochi.domain.words.WordRepository

class LevelContentProvider(
    private val kanaRepository: KanaRepository,
    private val kanjiRepository: KanjiRepository,
    private val wordRepository: WordRepository,
    private val scoreRepository: ScoreRepository
) {

    private val kanaPoolCache = mutableMapOf<String, List<String>>()

    fun getCharactersForLevel(levelKey: String, scoreType: ScoreManager.ScoreType = ScoreManager.ScoreType.RECOGNITION): List<String> {
        val lowerKey = levelKey.lowercase()
        return when {
            // Static / Special mappings
            lowerKey == "hiragana" -> kanaRepository.getKanaEntries(KanaType.HIRAGANA).map { it.character }
            lowerKey == "katakana" -> kanaRepository.getKanaEntries(KanaType.KATAKANA).map { it.character }
            lowerKey == "native_challenge" || lowerKey == "native challenge" -> kanjiRepository.getNativeKanji().map { it.character }
            lowerKey == "no_reading" || lowerKey == "no reading" -> kanjiRepository.getNoReadingKanji().map { it.character }
            lowerKey == "no_meaning" || lowerKey == "no meaning" -> kanjiRepository.getNoMeaningKanji().map { it.character }
            lowerKey == "user_custom_list" -> {
                val listName = when(scoreType) {
                    ScoreManager.ScoreType.RECOGNITION -> ScoreManager.RECOGNITION_LIST
                    ScoreManager.ScoreType.READING -> ScoreManager.READING_LIST
                    ScoreManager.ScoreType.WRITING -> ScoreManager.WRITING_LIST
                    ScoreManager.ScoreType.GRAMMAR -> ScoreManager.RECOGNITION_LIST
                }
                scoreRepository.getListItems(listName)
            }
            
            // Word lists
            lowerKey.contains("wordlist") -> wordRepository.getWordsForLevel(levelKey)
            
            // Default
            else -> kanjiRepository.getKanjiByLevel(levelKey).map { it.character }
        }
    }

    /**
     * Optimized: returns a unique pool of kana used in the words of a specific level.
     * Cached to speed up game initialization (Kana Drop, Simon, etc.)
     */
    suspend fun getKanaPoolForLevel(levelKey: String): List<String> {
        kanaPoolCache[levelKey]?.let { return it }

        val wordEntries = try {
            wordRepository.getWordEntriesForLevelSuspend(levelKey)
        } catch (e: Exception) {
            emptyList()
        }
        
        val pool = wordEntries.flatMap { word ->
            word.phonetics.map { it.toString() }
        }.distinct().filter { 
            it.isNotBlank() && it.first() !in " ./()[]+-*=_\"'!?#%&"
        }

        val finalPool = pool.ifEmpty { 
            // Fallback to basic hiragana if no pool found
            kanaRepository.getKanaEntries(KanaType.HIRAGANA).map { it.character } 
        }
        
        kanaPoolCache[levelKey] = finalPool
        return finalPool
    }
}

File: flutter_app/lib/services/level_content_provider.dart
import '../repositories/kana_repository.dart';
import '../repositories/dictionary_repository.dart';
import '../repositories/word_repository.dart';
import '../repositories/score_repository.dart';
import '../models/kana.dart';

class LevelContentProvider {
  final KanaRepository kanaRepo;
  final DictionaryRepository dictionaryRepo;
  final WordRepository wordRepo;
  final ScoreRepository scoreRepo;

  LevelContentProvider({
    required this.kanaRepo,
    required this.dictionaryRepo,
    required this.wordRepo,
    required this.scoreRepo,
  });

  Future<List<String>> getItemsForLevel(String levelKey, ScoreType type, String locale) async {
    final lowerKey = levelKey.toLowerCase();

    // 1. Cas spéciaux statiques
    if (lowerKey == "hiragana") {
      final entries = await kanaRepo.getKanaEntries(KanaType.hiragana);
      return entries.map((e) => e.character).toList();
    }
    if (lowerKey == "katakana") {
      final entries = await kanaRepo.getKanaEntries(KanaType.katakana);
      return entries.map((e) => e.character).toList();
    }

    // 2. Liste de révision personnalisée
    if (lowerKey == "user_custom_list") {
      final listName = _getListNameForType(type);
      return await scoreRepo.getListItems(listName);
    }

    // 3. Kanji & Défis techniques
    final allKanji = await dictionaryRepo.getFullDictionary(locale);

    if (lowerKey.contains("native_challenge")) {
      return allKanji.where((k) => k.levelIds.isEmpty && k.readings.isNotEmpty).map((k) => k.character).toList();
    }
    if (lowerKey.contains("no_reading")) {
      return allKanji.where((k) => k.levelIds.isEmpty && k.readings.isEmpty && k.meanings.isNotEmpty).map((k) => k.character).toList();
    }
    if (lowerKey.contains("no_meaning")) {
      return allKanji.where((k) => k.levelIds.isEmpty && k.meanings.isEmpty).map((k) => k.character).toList();
    }

    // 4. Par défaut : Filtrage par niveau JLPT/Ecole
    return allKanji.where((k) => k.levelIds.any((l) => l.toLowerCase() == lowerKey)).map((k) => k.character).toList();
  }

  String _getListNameForType(ScoreType type) {
    switch (type) {
      case ScoreType.recognition: return "Recognition_List";
      case ScoreType.reading: return "Reading_List";
      case ScoreType.writing: return "Writing_List";
      case ScoreType.grammar: return "Grammar_List";
    }
  }
}
